columnq/src/table/mod.rs

```diff
@@ -405,9 +406,14 @@ impl TableSource {
 
     pub fn parsed_uri(&self) -> Result<URIReference, ColumnQError> {
         match &self.io_source {
-            TableIoSource::Uri(uri) => URIReference::try_from(uri.as_str()).map_err(|_| {
-                ColumnQError::InvalidUri(format!("{uri}. Make sure it's URI encoded."))
-            }),
+            TableIoSource::Uri(uri) => {
+                let uri_string = match Url::parse(uri) {
+                    Ok(encoded_uri) => String::from(encoded_uri),
+                    // fallback
+                    Err(_) => String::from(uri),
+                };
+                URIReference::try_from(uri_string.as_str()).map_err(|e| { ColumnQError::InvalidUri(format!("{uri_string}: {e:?}")) })
```

Comment:
[TODO] fix borrow checker

```
error[E0515]: cannot return value referencing local variable `uri_string`
   --> columnq/src/table/mod.rs:415:17
    |
415 |                 URIReference::try_from(uri_string.as_str()).map_err(|e| { ColumnQError::InvalidUri(format!("{uri_string}: {e:?}")) })
    |                 ^^^^^^^^^^^^^^^^^^^^^^^-------------------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                 |                      |
    |                 |                      `uri_string` is borrowed here
    |                 returns a value referencing data owned by the current function

```

Reply:
This is because `Url::parse` creates a heap allocated string, which is not owned by any variable or struct outside of this function. So the returned URIReference will be holding an invalid memory reference once this function returns.

To fix this, we need to store the return value from `Url::parse`. How about we implement a custom deserializer for `TableIoSource::Uri` and change the enum variant type from `String` to `url::Url`? I think we might be able to get rid of the use of `uriparse::URIReference` entirely and use `url::Url` every where.

Reply:
> I think we might be able to get rid of the use of uriparse::URIReference entirely and use url::Url every where.

sounds good, let me try

Reply:
> This is because Url::parse creates a heap allocated string, which is not owned by any variable or struct outside of this function

ah, thanks for detail explainations